### ISSUE TITLE
possible xss attack in StandardUrlAttributePolicy

### ISSUE THREAD

**Issue by saaspeter** [OWASP/java-html-sanitizer#213]
if user input sting is: 

```html
<body>aaa bbb<a href=\"jav&#97script:alert(1)\">test1</a></body>
```

policy defined as this:

```java
PolicyFactory LINKS_RAW = (new HtmlPolicyBuilder()).allowElements("a").allowStandardUrlProtocols().allowAttributes("href","target").onElements("a").toFactory()
```

when check the "a" tag attribute href,  StandardUrlAttributePolicy will be used to the the href value, this method:  apply(String elementName, String attributeName, String value) will be called, but since there is '#' in the value, so it skipped the protocol check, and finally return the input href value, so even the value which not contain any protocol in "HTTP, https, mailto" , will not be filtered.

![image](https://user-images.githubusercontent.com/3772321/100342024-56e30900-3018-11eb-96c4-e6f9c8edf8bc.png)

is this a bug?

**Comment by yangbongsoo** [comment]
@saaspeter Hello. I am sanitizer user.
I tested what you mentioned.

```java
  @Test
  public void testIssue213() {
    PolicyFactory policyFactory = new HtmlPolicyBuilder()
            .allowElements("a")
            .allowStandardUrlProtocols()
            .allowAttributes("href", "target")
            .onElements("a")
            .toFactory();
    String sanitize = policyFactory.sanitize("<a href=\"jav&#97script:alert(1)\">test1");
    System.out.println(sanitize); // <a href="jav&amp;#97script:alert%281%29">test1</a>
```
the result string is `<a href="jav&amp;#97script:alert%281%29">test1</a>`
even if href attribute value is remained, that string can not possible attack.
sanitizer transform `&` to `&amp;`


![스크린샷 2020-11-30 오후 4 17 16](https://user-images.githubusercontent.com/7060510/100579647-e42fa180-3327-11eb-8f6c-133487f7362c.png)

**Comment by mikesamuel** [comment]
Thanks for the report.  Please do try to report bypasses as described in the issue template

>.   Please report **security vulnerabilities** via [OWASP's vulnerability rewards program](https://bugcrowd.com/owaspjavasanitizer).

I believe @yangbongsoo is correct.  `&#97` is missing a `;` so is not an HTML character reference, but if it were, it would be decoded to `'a'` before `apply` is called.

**Comment by saaspeter** [comment]
@yangbongsoo , you are right, for my cast the original string will be sanitized, but I take the original string if there is no attacked string in it, so when the StandardUrlProtocols failed the detection, I use the original string :( , what's more I don't want my string was sanitized if there is no risk in the input string.  Anyway, I think if the href is not a standard protocal url, StandardUrlProtocols validation should fail. 
for my case I wrote a new enhanced policy to replace StandardUrlProtocols.  
thanks

**Comment by jmanico** [comment]
Confirming the analysis above on the current release. `&#97` without a trailing `;` is not a character reference, and the `#` makes the value a relative URL, so it is kept with the `&` escaped and is inert. Add the `;` and it decodes to `javascript:`, and the attribute is dropped. `allowStandardUrlProtocols()` permits relative URLs by design.

One note on the workaround mentioned: falling back to the original input when the sanitizer changes it defeats the point, since the escaping is exactly what makes this case harmless. If the policy is too permissive for your use, a custom `AttributePolicy` on `href` is the right place to tighten it.

Closing as working as intended.
